apps/remote_control/test/lexical/remote_control/code_intelligence/variable_test.exs

```diff
@@ -0,0 +1,489 @@
+          end
+        ]
+        |> find_references(true)
+
+      assert decorate(doc, definition) =~ "def something(«param») do"
+      assert decorate(doc, reference) =~ "  «param»"
+    end
+
+    test "can be found via a usage" do
+      {:ok, [first, second, third], doc} =
+        ~q[
+          def something(param) do
+           y = param + 3
+           z = param + 4
+           param| + y + z
+          end
+        ]
+        |> find_references()
+
+      assert decorate(doc, first) =~ " y = «param» + 3"
+      assert decorate(doc, second) =~ " z = «param» + 4"
+      assert decorate(doc, third) =~ " «param» + y + z"
+    end
+
+    test "are found in a function body" do
+      {:ok, [first, second, third, fourth, fifth], doc} =
+        ~q[
+          def something(param|) do
+            x = param + param + 3
+            y = param + x
+            z = 10 + param
+            x + y + z + param
+          end
+        ]
+        |> find_references()
+
+      assert decorate(doc, first) =~ "  x = «param» + param + 3"
+      assert decorate(doc, second) =~ "  x = param + «param» + 3"
+      assert decorate(doc, third) =~ "  y = «param» + x"
+      assert decorate(doc, fourth) =~ "  z = 10 + «param»"
+      assert decorate(doc, fifth) =~ " x + y + z + «param»"
+    end
+
+    test "are constrained to their definition function" do
+      {:ok, [range], doc} =
+        ~q[
+          def something(param|) do
+            param
+          end
+
+          def other_fn(param) do
+            param + 1
+          end
+        ]
+        |> find_references()
+
+      assert decorate(doc, range) =~ "«param»"
+    end
+
+    test "are visible across blocks" do
+      {:ok, [first, second], doc} =
+        ~q[
+          def something(param|) do
+            if something() do
+              param + 1
+            else
+              param + 2
+            end
+          end
+        ]
+        |> find_references()
+
+      assert decorate(doc, first) =~ "  «param» + 1"
+      assert decorate(doc, second) =~ "  «param» + 2"
+    end
+
+    test "dont leak out of blocks" do
+      {:ok, [range], doc} =
+        ~q[
+          def something(param) do
+
+            if something() do
+              param| = 3
+              param + 1
+            end
+            param + 1
+          end
+        ]
+        |> find_references()
+
+      assert decorate(doc, range) =~ "«param»"
+    end
+
+    test "are found in the head of a case statement" do
+      {:ok, [range], doc} =
+        ~q[
+          def something(param|) do
+            case param do
+             _ -> :ok
+            end
+          end
+        ]
+        |> find_references()
+
+      assert decorate(doc, range) =~ "  case «param» do"
+    end
+
+    test "are constrained to a single arm of a case statement" do
+      {:ok, [range], doc} =
+        ~q[
+          def something(param) do
+            case param do
+             param| when is_number(param) -> param + 1
```

Comment:
There might be something odd here where this `is_number(params)` cannot be found, and similarly, when the cursor is at this position: `|param + `, it can't find the definition on the left either.

Reply:
Yep, we weren't handling guards properly. They have a very odd AST representation that needs to be handled specially in both the `def` cases and the anonymous function cases.